src/app/public/modules/text-editor/services/text-editor.service.ts

```diff
@@ -0,0 +1,501 @@
+    /* istanbul ignore next */
+    let fontSize = STYLE_STATE_DEFAULTS.fontSize.toString();
+    const selection = this.getCurrentSelection(id);
+    if (
+      selection &&
+      selection.anchorNode &&
+      selection.anchorNode.parentElement
+    ) {
+      let element = selection.anchorNode;
+      if (element.nodeType !== 1) {
+        element = element.parentElement;
+      }
+      const computedStyle = window.getComputedStyle(element as Element);
+      /* istanbul ignore else */
+      if (computedStyle) {
+        fontSize = computedStyle.getPropertyValue('font-size');
+      }
+    }
+    return fontSize;
+  }
+
+  private createObservers(id: string, element: HTMLIFrameElement): EditorSetting {
+    /* istanbul ignore next */
+    const documentEl = element.contentWindow ?
+      element.contentWindow.document :
+      element.contentDocument;
+
+    // Firefox bug where we need to open/close to cancel load so it doesn't overwrite attrs
+    documentEl.open();
+    documentEl.close();
+
+    const selectionObservable = new Subject();
+    const selectionListener = () => selectionObservable.next();
+    const clickObservable = new Subject();
+    const clickListener = () => clickObservable.next();
+    const pasteListener = this.getPasteOverride(id);
+
+    documentEl.addEventListener('selectionchange', selectionListener);
+    documentEl.addEventListener('input', selectionListener);
+    documentEl.addEventListener('mousedown', clickListener);
+    documentEl.body.addEventListener('paste', pasteListener);
+    return {
+      iframeElementRef: element,
+      selectionListener: selectionListener,
+      clickListener: clickListener,
+      pasteListener: pasteListener,
+      selectionChangeObservable: selectionObservable,
+      clickObservable: clickObservable,
+      commandChangeObservable: new Subject()
+    };
+  }
+
+  private static removeObservers(setting: EditorSetting): void {
+    /* istanbul ignore next */
+    const documentEl = setting.iframeElementRef.contentWindow ?
+       setting.iframeElementRef.contentWindow.document :
+       setting.iframeElementRef.contentDocument;
+    setting.selectionChangeObservable.complete();
+    setting.clickObservable.complete();
+    setting.commandChangeObservable.complete();
+    documentEl.removeEventListener('selectionchange', setting.selectionListener);
+    documentEl.removeEventListener('input', setting.selectionListener);
+    documentEl.removeEventListener('mousedown', setting.clickListener);
+    documentEl.body.removeEventListener('paste', setting.pasteListener);
+  }
+
+  /* istanbul ignore next */
+  private getPasteOverride(id: string): (e: ClipboardEvent) => void {
+    return (e: ClipboardEvent): void => {
+      e.preventDefault();
+      const text = e.clipboardData.getData('text/plain');
+      this.execCommand(id, { command: 'insertHTML', value: text });
+    };
+  }
+
+  private static getColor(documentEl: Document, selector: string): string {
```

Comment:
Why are some of these static?

Reply:
It seems a little odd to me to have a service that mixes static and non-static methods. Not the end of the world but not something we do very often.

Reply:
The contributor deliberately made this decision because they didn't need an instance reference.